### PR TITLE
Add fork_at_current_lsn function which creates branch at current LSN

### PR DESCRIPTION
**Pull request by knizhnik**
There are some test failures caused by wrong assumption that all changes made before `create_branch` are applied at server before branch creation and so included in the branch.

I added `fork_at_current_lsn` function which extracts current LSN and pass it to `create_branch` function.

### CODE REVIEW

**Comment by hlinnaka** [comment]
relevant github issue: https://github.com/neondatabase/neon/issues/2063

File: test_runner/fixtures/neon_fixtures.py [hunk 1]
```diff
@@ -2637,3 +2637,15 @@ def wait_for_last_flush_lsn(env: NeonEnv, pg: Postgres, tenant: uuid.UUID, timel
...
+
+
+def fork_at_current_lsn(
+    env: NeonEnv,
+    pg: Postgres,
+    new_branch_name: str = DEFAULT_BRANCH_NAME,
+    ancestor_branch_name: Optional[str] = None,
+    tenant_id: Optional[uuid.UUID] = None,
+) -> uuid.UUID:
+    """Create new branch for current tenant/timeline on the last LSN. Before branch creation this function enforce that LSN is replayed by pagserver"""
```
Comment (hlinnaka) [inline test_runner/fixtures/neon_fixtures.py:2649]:
```suggestion
    """
    Create new branch at the last LSN of an existing branch.
    The "last LSN" is taken from the given Postgres instance. The pageserver will wait for all the
    the WAL up to that LSN to arrive in the pageserver before creating the branch.
    """
```

File: test_runner/fixtures/neon_fixtures.py [hunk 2]
```diff
@@ -2637,3 +2637,15 @@ def wait_for_last_flush_lsn(env: NeonEnv, pg: Postgres, tenant: uuid.UUID, timel
     """Wait for pageserver to catch up the latest flush LSN"""
     last_flush_lsn = lsn_from_hex(pg.safe_psql("SELECT pg_current_wal_flush_lsn()")[0][0])
     wait_for_last_record_lsn(env.pageserver.http_client(), tenant, timeline, last_flush_lsn)
+
+
+def fork_at_current_lsn(
+    env: NeonEnv,
+    pg: Postgres,
+    new_branch_name: str = DEFAULT_BRANCH_NAME,
+    ancestor_branch_name: Optional[str] = None,
```
Comment (hlinnaka) [inline test_runner/fixtures/neon_fixtures.py:2646]: I think `ancestor_branch_name` be mandatory. This doesn't really make sense without an ancestor branch.

```suggestion
    ancestor_branch_name: str,
```

File: test_runner/fixtures/neon_fixtures.py [hunk 3]
```diff
@@ -2637,3 +2637,15 @@ def wait_for_last_flush_lsn(env: NeonEnv, pg: Postgres, tenant: uuid.UUID, timel
     """Wait for pageserver to catch up the latest flush LSN"""
     last_flush_lsn = lsn_from_hex(pg.safe_psql("SELECT pg_current_wal_flush_lsn()")[0][0])
     wait_for_last_record_lsn(env.pageserver.http_client(), tenant, timeline, last_flush_lsn)
+
+
+def fork_at_current_lsn(
+    env: NeonEnv,
+    pg: Postgres,
+    new_branch_name: str = DEFAULT_BRANCH_NAME,
```
Comment (hlinnaka) [inline test_runner/fixtures/neon_fixtures.py:2645]: I think `new_branch_ name` should be mandatory too.

```suggestion
    new_branch_name: str,
```

**Review by hlinnaka — APPROVED** [review]
I think there are a few more tests that should use the new function instead of `create_branch`.

**Comment by knizhnik** [comment]
> I think there are a few more tests that should use the new function instead of `create_branch`.

I have inspected all placed where `create_branch` is called in our tests. In most cases it is called at startup when database is empty, so there is no need to wait until data s replayed by page server.
And in `test_branching.py` I have originally replaced `create_branch` with `fork_at_current_lsn` but later has to revert my change back. The reason is that test is using small `gc_hiorizon` and attempt to create fork at LSN taken from compute node cause "LSN {} is earlier than latest GC horizon {} (we might've already garbage collected needed data)" error.

**Pull request merged**